tutorials/circuits_advanced/qv_tools.py

```diff
@@ -0,0 +1,146 @@
+import matplotlib.pyplot as plt
+import numpy as np
+
+from qiskit.quantum_info import Statevector
+from qiskit.circuit import QuantumRegister
+from qiskit.circuit.library import XGate
+from qiskit.transpiler import PassManager, Layout
+from qiskit.transpiler.passes import (BIPMapping,
+                                      Depth, 
+                                      FixedPoint,
+                                      Collect2qBlocks, 
+                                      ConsolidateBlocks,
+                                      UnitarySynthesis, 
+                                      Optimize1qGatesDecomposition, 
+                                      CommutativeCancellation, 
+                                      SetLayout, 
+                                      FullAncillaAllocation, 
+                                      EnlargeWithAncilla, 
+                                      ApplyLayout,
+                                      ALAPSchedule,
+                                      DynamicalDecoupling)
+
+# a custom passmanager that does the compilation of QV circuits
+def qv_passmanager(basis_gates, coupling_map, qubit_subset, backend_props,
+                   instruction_durations, synthesis_fidelity, pulse_optimize):
+    
+    def _repeat_condition(property_set):
+        return not property_set['depth_fixed_point']
+    
+    sub_coupling_map = coupling_map.reduce(qubit_subset)
+    qr = QuantumRegister(size=len(qubit_subset), name='q')
+    layout = Layout.from_intlist(sorted(qubit_subset), qr)
+
+    _map = [BIPMapping(sub_coupling_map, objective='depth', time_limit=100)]
+
+    _embed = [SetLayout(layout),
+              FullAncillaAllocation(coupling_map),
+              EnlargeWithAncilla(),
+              ApplyLayout()]
+
+    _check_depth = [Depth(),
+                    FixedPoint('depth')]
+
+    _optimize = [
+        Collect2qBlocks(),
+        ConsolidateBlocks(basis_gates=basis_gates),
+        UnitarySynthesis(basis_gates, synthesis_fidelity, coupling_map, backend_props,
+                         pulse_optimize=True, natural_direction=True),
+        Optimize1qGatesDecomposition(basis_gates)
+    ]
+
+    _schedule = [
+        ALAPSchedule(instruction_durations),
+        DynamicalDecoupling(instruction_durations, dd_sequence=[XGate(), XGate()])
+    ]
+
+    pm = PassManager()
+    pm.append(_map)                        # map to a subset of qubits, by choosing layout and inserting swaps
+    pm.append(_embed)                      # embed the mapped circuit onto the larger device coupling map
+    pm.append(_check_depth + _optimize,
+              do_while=_repeat_condition)  # translate to & optimize over hardware native gates
+    pm.append(_schedule)                   # dynamical decoupling on an as-late-as-possible schedule
+    return pm
```

Comment:
Is there a reason to not put this in the tutorial text itself? The pass manager visualization is kind of hard to read and just having the code in the notebook is probably useful.

Reply:
done